language-support/ts/daml-types/index.ts

```diff
@@ -329,10 +356,32 @@ class OptionalWorker<T> implements Serializable<Optional<T>> {
         jtv.constant<[]>([]),
         jtv.tuple([payloadInnerDecoder]),
       ) as jtv.Decoder<OptionalInner<T>>;
+      this.encode = (o: Optional<T>): unknown => {
+        // We assume that we start with a well-formed value (that's what the
+        // types are for). Therefore, if we have an explicit null, we are at
+        // the "top level" of the tower of optionals and we can just send null
+        // over the wire.
+        if (o === null) {
+          return null;
+        } else {
+          // If we do not have null, and we are in the case of a nested
+          // optional, then we MUST have a list. At this point all we want is
+          // to recursively apply the decoders; they should all go through this
+          // branch now, except the final one.
```

Comment:
Not sure what you mean by “the final one”.

Reply:
If you have an `Optional<Optional<Optional<Int>>>` and you actually have a value, you'll have something like `[[[4]]]`. The first two `encode` calls will have type `Optional<Optional<?>>`, so they go through this branch (that's what the outer `if` tests on here). The last `Optional.encode` that gets called is for a type `Optional<Int>`, so it goes through the other branch of the outer `if` because its type argument is not itself an optional. This only happens if we do have a value eventually; if we have a nested `None` then all calls go through the first branch of the `if`.

Reply:
Perhaps “innermost” instead of “final” would be clearer?

Reply:
I've tried to make the comments clearer.